proposals/unsafe-evolution.md

```diff
@@ -0,0 +1,336 @@
+
+### Local functions/lambda safe contexts
+
+Right now `unsafe` on a method body is lexically scoped. Any nested local functions or lambdas inherit this, and their bodies are in a memory unsafe context. Is this behavior that we want to keep in
+the language? Note that if we do keep `unsafe` as the modifier used to expose that the caller must be unsafe, this could then have impacts on the signature of the method. As currently proposed, nested
+anonymous and local functions do not keep the unsafe context of their containing member.
+
+### Lambda/method group conversion to safe delegate types
+
+Is conversion of a lambda or method group marked `unsafe` to a non-unsafe delegate type permitted without warning or error in an `unsafe` context? If we don't do this, then it could be fairly painful
+for various parts of the ecosystem, particularly any enumerables that are passed through LINQ queries.
+
+### `stackalloc` as initialized
+
+Today, [the spec](https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#12822-stack-allocation) always considers `stackalloc` memory as uninitialized, and says that the contents
+are undefined unless manually cleared or assigned. Do we consider this a spec bug, or do we need to change what we consider `unsafe` for `stackalloc` purposes?
+
+### `unsafe` expressions
+
+Other languages with more comprehensive `unsafe` features have added `unsafe` as an expression, to enable improved user ergonomics and allow authors to more precisely limit where `unsafe` is used. Is this
+something that we want to have in C#? Consider an inline call to an `unsafe` member that handles the safety directly: right now, the author would either need to wrap the entire statement in an `unsafe`
+block, expanding the scope of the `unsafe` context, or they would need to break out the inner function call into an intermediate variable.
+
+```cs
+extern int Add(int i1, int i2); // Some fancy extern addition function
+
+// Code I want to write:
+Console.WriteLine(unsafe(Add(1, 2)));
+
+// Code I have to write option 1, unsafe context unnecessary includes the WriteLine call
+unsafe
+{
+    Console.WriteLine(Add(1, 2));
+}
+
+// Code I have to write option 2, very verbose and harder to read:
+int result;
+unsafe
+{
+    result = Add(1, 2);
+}
+Console.WriteLine(result);
+```
+
+### `unsafe` fields
+
+Today, no proposal is made around `unsafe` on a field. We may need to add it though, such that any read from or write to a field marked as `unsafe` must be in an `unsafe` context. This would
+enable us to better annotate the concerns around code such as:
+
+```cs
+class SafeWrapper
+{
+     internal byte* _p;
+
+     public void DoStuff()
+     {
+            unsafe
+            {
+                  // ... validate that the object state is good ...
+                  // ... perform operation with _p .... 
+            }
+     }
+
+}
+
+// Elsewhere in safe code:
+void M(SafeWrapper w)
+{
+     w._p = stackalloc byte[10];
```

Comment:
I'm not sure I see the benefit of marking the field itself unsafe. If it's dereferenced, that would be unsafe anyway. If it's manipulated via other APIs of SafeWrapper, they should be marked as `unsafe`, otherwise they declare they can handle anything in the field safely.

Reply:
In this case, the field exposed through the program. So anything that manipulates it needs to make sure that the pointer obeys the expectations of `SafeWrapper`.